### ISSUE TITLE
prototype = new Error() call causes problems

### ISSUE THREAD

**Issue by oliversalzburg** [KyleAMathews/superagent-bluebird-promise#25]
Not quite sure why, but it's apparently an issue:

```
Error
    at Object.<anonymous> (/usr/lib/node_modules/cli/node_modules/superagent-bluebird-promise/index.js:16:36)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.<anonymous> (/usr/lib/node_modules/cli/lib/install.js:15:15)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at attemptRequire (/usr/lib/node_modules/cli/node_modules/attempt-require/index.js:6:20)
    at Object.<anonymous> (/usr/lib/node_modules/cli/bin/cli.js:39:15)
```

**Comment by KyleAMathews** [comment]
Could you provide more details about how this error happened? Errors are supposed to be thrown if in fact an error happens.

**Comment by oliversalzburg** [comment]
Indeed. I have no idea why that part of the code was causing a problem, I just thought I'd mention it. I'm now simply using promisification anyway.

**Comment by KyleAMathews** [comment]
Was it always erroring? I.e on every request?

**Comment by oliversalzburg** [comment]
@KyleAMathews Yes, I only tried it for a couple of times. Maybe I wasn't using it right either. I don't understand why a simple `new Error()` caused that.

Maybe it is because I'm using [`attempt-require`](https://github.com/oliversalzburg/try-require) in my stack or something else I'm not seeing.

One theory I had was that there was an actual error happening and the stack is just wrong, because it is generated when you call `new Error()`.

In our own codebase, we use the following pattern to extend `Error`:
```js
function UserError( message ) {
	//noinspection JSUnresolvedFunction
	Error.captureStackTrace( this, this.constructor );
	this.name = this.constructor.name;
	this.message = message || "Bad request";
}

util.inherits( UserError, Error );
```

This is somewhat node specific though.

While writing this, it occurs to me that this actually sounds like a reasonable explanation.

**Comment by spudly** [comment]
I'm running into this as well.

@oliversalzburg, You're exactly right.

A new stack trace is created when the module is parsed and executed. Then every time a real error is thrown, it swallows the real error's stack trace and gives you the wrong stack trace.

**Comment by KyleAMathews** [comment]
@oliversalzburg @spudly could one of you put together a PR for this?

**Comment by spudly** [comment]
I'll work on it.

**Comment by KyleAMathews** [comment]
@spudly thanks!

**Comment by spudly** [comment]
TravisCI finally ran the build and it passed. Please merge & npm publish a new version when you can.

**Comment by KyleAMathews** [comment]
LGTM, @oliversalzburg want to take a look at it as well?

**Comment by oliversalzburg** [comment]
@KyleAMathews Looks fine. I guess it just loses the ability to type-check on the error.

**Comment by spudly** [comment]
Right. You can still check the error's .name property, but it's just a plain instance of Error now. I could have tried to preserve the subclass logic and just improved it, but that gets tricky when dealing with older browsers.

**Comment by spudly** [comment]
Might want to do a minor or major version bump for this because of that little change...

**Comment by KyleAMathews** [comment]
Yeah, this does sound like it'd need a major bump. We should probably also add a CHANGELOG as well. @markdalgleish thoughts?

**Comment by KyleAMathews** [comment]
@spudly also which browsers specifically have trouble with the old custom error?

**Comment by oliversalzburg** [comment]
I'm just going to leave this here :) https://github.com/ralt/newerror

**Comment by spudly** [comment]
What I meant was that certain features that I would have used to improve it are not supported by all browsers (specifically, Error.captureStackTrace, which is only available in V8 environments such as node and chrome).

I'm all for using that newerror lib if you want. Looks like a decent abstraction.

One issue though is that the code currently uses the original error as error.message, which seems pretty strange to me. Seems like in that case we ought to just reject with the original error, but then it wouldn't be a SuperagentPromiseError anymore so it might be harder to type-check.

**Comment by KyleAMathews** [comment]
So there are two places errors are generated. One is in our control when a status code is greater than 400. The other comes from the browser / other sources when the ajax request fails for some reason.

I think it makes sense to get rid of our custom Error and just set the error name.

This would require a major bump but numbers are cheap so I'm cool with that. We should definitely add a CHANGELOG to communicate things.

@markdalgleish @jordansexton @kumatch and others who know more about this and have helped out in the past, any thoughts? Does this plan make sense?

**Comment by KyleAMathews** [comment]
Also we wouldn't need the newerror library as we wouldn't have our own Error subclass anymore.

**Comment by spudly** [comment]
Doesn't seem like anyone has any objections... Can we go ahead and merge this in?

**Comment by jordansexton** [comment]
I actually do have an objection to this. bluebird's [`catch`](https://github.com/petkaantonov/bluebird/blob/master/API.md#catchfunction-errorclassfunction-predicate-function-handler---promise) wants a function or error class, not a string. That's means you'd have to add a handler for each you want to catch. Since this library is meant to work well with bluebird, this is the primary reason makes sense to expose a library-specific error type.

What I might suggest is actually catching browser errors and throwing the `SuperAgentPromiseError` with the original attached. This way, userland code can always just `catch` it and then examine it to see the cause if needed.

**Comment by spudly** [comment]
What do you guys think about bringing back the SuperAgentPromiseError, but improving it such that when it wraps around another error the stack property contains both stack traces? The stack trace would look something like this:

```
ERROR SuperAgentPromiseError Bad Request
  at /path/to/some/js/file.js:123:456
  at /path/to/some/other/js/file.js:123:456
...
Original Error: Error: cannot GET http://some.domain/path/to/some/failing/script (500)
  at /path/to/some/js/file.js:123:456
  at /path/to/some/other/js/file.js:123:456
```

**Comment by KyleAMathews** [comment]
@spudly that seems sensible to me.

**Comment by spudly** [comment]
All,

I've update the pull request (#26). Please have a look.

**Comment by spudly** [comment]
Is this going to get merged in?

**Comment by KyleAMathews** [comment]
Sorry, was waiting for more comments then forgot about it. @oliversalzburg / @jordansexton LGTM?

**Comment by KyleAMathews** [comment]
@spudly does `Object.defineProperty` as you're using it work on IE8? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty

Not very familiar with this call.

**Comment by jordansexton** [comment]
@KyleAMathews Yeah, this won't work in IE8 (no getters, no real workaround). Could be wrapped in a safety clause to check for `Object.defineProperty`

The implementation looks good to me. My only quibble is that I'm not a huge fan of the name of the `childError` property as it implies the existence of a hierarchy or grouping upon the error. What do you think about just `cause` as is referenced in the stack trace?

**Comment by spudly** [comment]
I'll wrap the call in an if statement. It works if you use es5-shim and es5-sham so I'm used to using those functions without worrying about IE8, but I know that's not a safe assumption to make for a general-use library like this.

Not sure I like "cause" because it is not clear from the variable name what the variable contains. 
* originalError
* sourceError
* wrappedError

Let me know what you think. I'm not married to any of these variable names and will happily just call it "cause" if that's what you guys want.

**Comment by jordansexton** [comment]
+1 for originalError

**Comment by spudly** [comment]
Just got back from vacation... I made the changes requested:

* wrapped Object.defineProperty in an if statement for IE8 compatibility
* changed childError to originalError

**Comment by ce07c3** [comment]
Let's merge this.

**Comment by KyleAMathews** [comment]
@ce07c3 thanks for the reminder! @spudly's last update slipped away from me. Merged and published 2.1.0. Thanks everyone!
